Ash.FrontEnd/src/js/global/components/Fido.jsx

```diff
@@ -0,0 +1,293 @@
+ * @prop {object} query - object containing query string key-value pairs
+ * @prop {object} payload - json payload for request
+ * @prop {object} options - other options [optional]
+ * @prop {string} options.method - HTTP method type (case-insensitive) [optional]
+ * @prop {boolean} options.callOnMount - flag for whether the fetch should be
+ *    performed on mount [optional]
+ * @prop {boolean} options.preservePreviousData - flag for whether the fetch should
+ *    preserve the previously returned data set when a new call is initiated [optional]
+ */
+const defaultFetchConfig = {
+  path: '',
+  query: null,
+  payload: null,
+  options: {
+    method: 'get',
+    callOnMount: true,
+    preservePreviousData: true,
+  },
+};
+
+/**
+ * Fetch state object default object shape
+ * @prop {boolean} inFlight - is the request awaiting a reply?
+ * @prop {boolean} success - did we receive a good reply from the server?
+ * @prop {boolean} fail - did the request fail or did the server return a bad response?
+ * @prop {object} data - response payload of the successful request
+ * @prop {object} error - js error object
+ * @prop {Function} call - function to initiate a fetch
+ */
+const initialFetchState = {
+  inFlight: false,
+  success: false,
+  fail: false,
+  data: null,
+  error: null,
+  call: null,
+};
+
+/**
+ * Generate fetch state objects
+ * @param {object} fetchItems - Object containing named config objects
+ * @param {object} callFunc - "call" function
+ * @returns {object} Fetch config objects, keyed by config name, composed of
+ *    the default state above and the "call" function
+ */
+function createInitFetchState(fetchItems, callFunc) {
+  return Object.keys(fetchItems).reduce((initState, key) => {
+    initState[key] = {
+      ...initialFetchState,
+      call: callFunc(key),
+    };
+
+    return initState;
+  }, {});
+}
+
+/**
+ * Fetch React Component for retrieving data from a server using the *fetch* API
+ * Follows the render props pattern for wrapping children
+ * @prop {object} fetchItems - Collection of configurations for fetch calls.
+ *    Will generate new state for any new fetch items received, but will not discard any
+ *    previous fetch item states until being unmounted.
+ */
+class Fido extends React.Component {
+  constructor(props) {
+    super(props);
+
+    const { fetchItems } = props;
+    this.state = createInitFetchState(fetchItems, this.onCall);
+    this.aborts = {};
+  }
+
+  componentDidMount() {
+    const configs = this.getConfigs();
+
+    // Call any fetch items with the callOnMount option set
+    Object.entries(configs).forEach(([key, config]) => {
+      if (config.options.callOnMount) {
+        this.dispatchFetch(key, config);
+      }
+    });
+  }
+
+  componentDidUpdate(prevProps) {
+    const { fetchItems } = this.props;
+
+    if (fetchItems !== prevProps.fetchItems) {
+      // Generate default state for any new fetch items (don't mess with existing ones)
```

Comment:
Are you just ignoring cases where they may change the options of a particular request? (Not 100% yet if this is an issue for you, but I know mine had options like running every render or not/etc)

Reply:
So, this has to do with the fetch state, not the fetch configurations.  Those are a part of a different object.  The config is actually generated from a memoized function that takes in the `fetchItems` prop.

The idea here is that if the fetchItems prop (read fetch config input) changed and there are new keys, a new initial state object is created for those new fetchItem keys.  I maybe should try to improve the names of these things.

Reply:
Ahh okay makes sense